### ISSUE TITLE
**How to Improve Semantic Quality of Root Embeddings**

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#341]
## Current Architecture Analysis

**What we have:**
- **Model**: Simple embedding layer (11,121 roots × 64d = 712KB params)
- **Training**: Graded similarity targets from ReVo + Ekzercaro co-occurrence
- **Loss**: MSE regression + margin loss for separation
- **Performance**: 88% correlation on training set, BUT lacks semantic signal for retrieval

**Current training data sources:**
1. ✅ **Ekzercaro co-occurrence** (Zamenhof's examples) - graded 0.3-0.9
2. ✅ **ReVo Jaccard** (definition overlap) - graded 0.4-0.8
3. ✅ **ReVo semantic relations** (synonyms, antonyms, hypernyms) - bonus weight

**Key insight**: Model is trained for **similarity consistency** (words with similar definitions should be similar), NOT for **semantic retrieval** (words related to a concept should cluster together).

## Why Current Embeddings Fail at Retrieval

1. **Jaccard-based training is weak**:
   - "Fundamento" and "Esperanto" may share few definition words
   - Their embeddings could be far apart even though semantically related
   - Example: "Fundamento" defined as "fundamenta libro de Esperanto" (1887, 1905)
   - But "1905" and "Fundamento" don't co-occur in definitions → no training signal

2. **No document-level context**:
   - Current training uses sentence co-occurrence (Ekzercaro)
   - But doesn't use 4.3M corpus documents
   - Missing: "Words that appear in similar contexts should be similar"

3. **Function word filtering helps morphology, hurts retrieval**:
   - Filtering "est", "far", "hav" prevents embedding collapse ✓
   - But also loses important semantic context for retrieval

## Solutions: NO Architecture Change Needed!

**Good news**: Current architecture is fine. We just need better training data.

### Option 1: Add Corpus Context (RECOMMENDED)

**Use Skip-Gram / CBOW objectives on 4.3M corpus**:

```python
# For each sentence in corpus:
#   For each word w:
#     Predict context words from w (skip-gram)
#     OR predict w from context words (CBOW)
#
# This learns: words in similar contexts → similar embeddings
```

**Implementation**:
1. Multi-task loss: `0.5 * similarity_loss + 0.5 * skipgram_loss`
2. Keep existing ReVo + Ekzercaro training (it works!)
3. Add corpus context as additional signal
4. Train for more epochs (currently 45 epochs)

**Expected improvement**: 12% → 25-30% retrieval accuracy

### Option 2: Use Q&A Pairs as Supervision

**Create training pairs from Q&A data**:
- Question: "Kiam aperis la Fundamento?"
- Answer doc contains: "Fundamento", "1905", "aperis"
- Training signal: These roots should be semantically related

**Implementation**:
1. Extract question roots: [kiam, aper, fundament]
2. Extract answer roots: [fundament, 1905, aper, esperant]
3. Create positive pairs: (fundament, 1905), (fundament, aper)
4. Create negative pairs: sample random roots
5. Train with contrastive loss

**Blocker**: Need answerable Q&A pairs (see Task #66 - audit corpus coverage)

### Option 3: Hybrid Approach (BEST)

**Combine both**:
1. **Base training**: ReVo + Ekzercaro (current) - 40% weight
2. **Corpus context**: Skip-gram on 4.3M docs - 40% weight
3. **Task-specific**: Q&A pairs where available - 20% weight

Multi-task loss balances all objectives.

## Do We Need to Retrain from Scratch?

**NO!** We can **fine-tune** the existing model:

```bash
# Resume from checkpoint
python scripts/training/train_root_embeddings.py \
  --resume models/root_embeddings/best_model.pt \
  --add-corpus-context \
  --epochs 20  # Additional epochs
```

**Advantages**:
- Keep existing semantic knowledge (ReVo relations)
- Add corpus context signal
- Faster than training from scratch (starts from good initialization)

**Disadvantages**:
- May not be as optimal as full retraining
- But much faster to experiment

## Recommended Approach

**Phase 1: Quick experiment (2-3 hours)**
1. Create corpus context training pairs from 4.3M docs
2. Fine-tune existing model with multi-task loss
3. Rebuild HNSW index with new embeddings
4. Re-run benchmark → expect 20-25% accuracy

**Phase 2: Full retraining (1-2 days if needed)**
1. If fine-tuning helps, retrain from scratch
2. Optimize hyperparameters (learning rate, loss weights)
3. Train for 100+ epochs
4. Target: 30-40% retrieval accuracy

**Phase 3: Task-specific tuning**
1. After corpus coverage audit (Task #66)
2. Add Q&A supervision for answerable questions
3. Target: 50-60% accuracy

## Key Insight

**We DON'T need to change architecture**. The simple embedding layer is fine. We just need to:
1. ✅ Keep ReVo + Ekzercaro training (handles linguistic similarity)
2. ➕ Add corpus context (handles semantic clustering)
3. ➕ Add Q&A supervision (handles retrieval task)

All can be done with multi-task training on existing model.

**Comment by marctjones** [comment]
Phase 6 cleanup - see CURRENT_PLAN.md for active work
